### PR TITLE
Add lmdeploy example for model upload

### PR DESCRIPTION
**Pull request by phatvo9**

### CODE REVIEW

File: models/model_upload/llms/lmdeploy-llama-3_2-1b-instruct/requirements.txt [hunk 1]
```diff
@@ -0,0 +1,15 @@
+torch==2.4
+tokenizers
+transformers
```
Comment (luv-bansal) [inline models/model_upload/llms/lmdeploy-llama-3_2-1b-instruct/requirements.txt:3]: We should also add dependencies versions with the package name for each of depencencies

File: models/model_upload/llms/lmdeploy-llama-3_2-1b-instruct/1/model.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+from clarifai.runners.models.model_runner import ModelRunner
+from clarifai.utils.logging import logger
+from clarifai_grpc.grpc.api import resources_pb2, service_pb2
+from clarifai_grpc.grpc.api.status import status_code_pb2
+from google.protobuf import json_format
+from lmdeploy import pipeline, GenerationConfig, TurbomindEngineConfig
+from transformers import AutoTokenizer
+
+class MyRunner(ModelRunner):
+  """A custom runner that loads the model and generates text using vLLM Inference.
```
Comment (luv-bansal) [inline models/model_upload/llms/lmdeploy-llama-3_2-1b-instruct/1/model.py:16]: The comment need to be updated for lmdeploy

File: models/model_upload/llms/lmdeploy-llama-3_2-1b-instruct/1/model.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+  """
+
+  def load_model(self):
+    """Load the model here and start the vllm server."""
```
Comment (luv-bansal) [inline models/model_upload/llms/lmdeploy-llama-3_2-1b-instruct/1/model.py:20]: same, update comment for lmdeploy

File: models/model_upload/llms/lmdeploy-llama-3_2-1b-instruct/1/model.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    os.path.join(os.path.dirname(__file__))
+    # if checkpoints section is in config.yaml file then checkpoints will be downloaded at this path during model upload time.
+    checkpoints = os.path.join(os.path.dirname(__file__), "checkpoints")
+    # Note that Llama3.2-1B-Inst is not supported by turbomind yet, lmdeploy will auto switch to pytorch engine
```
Comment (luv-bansal) [inline models/model_upload/llms/lmdeploy-llama-3_2-1b-instruct/1/model.py:24]: I think we should add model example that `lmdeploy` but we also should include this comment here for knowledge purpose

File: models/model_upload/llms/lmdeploy-llama-3_2-1b-instruct/config.yaml [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+inference_compute_info:
+  cpu_limit: "2"
+  cpu_memory: "12Gi"
+  num_accelerators: 1
+  accelerator_type: ["NVIDIA-A10G"]
+  accelerator_memory: "24Gi"
+
+checkpoints:
+  type: "huggingface"
+  repo_id: "meta-llama/Llama-3.2-1B-Instruct"
```
Comment (luv-bansal) [inline models/model_upload/llms/lmdeploy-llama-3_2-1b-instruct/config.yaml:21]: Update this with LLM that lmdeploy supports for most efficiency of lmdeploy

**Review by luv-bansal — CHANGES_REQUESTED** [review]
Added some suggestions, I'll test this PR by uploading model on Model, but PR looks good to me

File: models/model_upload/llms/lmdeploy-llama-3_2-1b-instruct/1/model.py (hunk 4 above)
Comment (phatvo9) [reply models/model_upload/llms/lmdeploy-llama-3_2-1b-instruct/1/model.py]: The latest lmdeploy version now supports this model with the turbomind backend. 
I'm considering keeping a consistent model ID across all inference frameworks to easily compare performance differences. The PyTorch backend, being one of two in lmdeploy, is well-optimized as well.

File: models/model_upload/llms/lmdeploy-llama-3_2-1b-instruct/requirements.txt (hunk 1 above)
Comment (phatvo9) [reply models/model_upload/llms/lmdeploy-llama-3_2-1b-instruct/requirements.txt]: Added version specifications for key packages: transformers, protobuf, torch, etc. The remaining packages will be managed by pip to align with the specified versions.

File: models/model_upload/llms/lmdeploy-llama-3_2-1b-instruct/requirements.txt
Comment (luv-bansal) [inline models/model_upload/llms/lmdeploy-llama-3_2-1b-instruct/requirements.txt:1]: currently `pip install -r requirements.txt` is failing

Comment (luv-bansal) [reply models/model_upload/llms/lmdeploy-llama-3_2-1b-instruct/requirements.txt:1]: added dependencies version here and it's working
```
torch==2.4.0
tokenizers==0.20.2
transformers==4.46.2
accelerate==0.34.2
optimum==1.23.3
xformers==0.0.27.post2
einops==0.8.0
requests==2.32.2
packaging
ninja
protobuf==3.20.0

openai==1.51.0
qwen-vl-utils
lmdeploy==0.6.3
```

File: models/model_upload/llms/lmdeploy-llama-3_2-1b-instruct/config.yaml [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+build_info:
+  python_version: "3.10"
+
+inference_compute_info:
+  cpu_limit: "2"
+  cpu_memory: "12Gi"
+  num_accelerators: 1
+  accelerator_type: ["NVIDIA-A10G"]
+  accelerator_memory: "24Gi"
```
Comment (luv-bansal) [inline models/model_upload/llms/lmdeploy-llama-3_2-1b-instruct/config.yaml:17]: We can reduce this for 1billion parameter model

**Review by luv-bansal — COMMENTED** [review]
I've uploaded the model upload with all these changes to prod and predictions are successful.
After these changes we can merge this PR

**Review by luv-bansal — APPROVED** [review]

**Pull request merged**